### PR TITLE
Refactor config

### PR DESCRIPTION
**Pull request by IntelOrca**
Maybe not as elegant, but if we want to use `std::string` in the config models eventually - the memcpy won't work and will require a more engineered solution.

Again removing more uses of RWops.

### CODE REVIEW

**Comment by janisozaur** [comment]
```
/home/travis/build/OpenRCT2/OpenRCT2/test/testpaint/compat.c:17:29: fatal error: openrct2/config.h: No such file or directory
```

```
In file included from /home/travis/build/OpenRCT2/OpenRCT2/src/openrct2/paint/sprite/peep.c:18:0:
/home/travis/build/OpenRCT2/OpenRCT2/src/openrct2/paint/sprite/../../drawing/lightfx.h:69:9: error: ‘struct SDL_Texture’ declared inside parameter list will not be visible outside of this definition or declaration [-Werror]
  struct SDL_Texture * texture,
         ^~~~~~~~~~~
```

File: src/openrct2/config/Config.cpp [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+    static auto Enum_MeasurementFormat = ConfigEnum<sint32>(
+    {
+        ConfigEnumEntry<sint32>("IMPERIAL", MEASUREMENT_FORMAT_IMPERIAL),
+        ConfigEnumEntry<sint32>("METRIC", MEASUREMENT_FORMAT_METRIC),
+        ConfigEnumEntry<sint32>("SI", MEASUREMENT_FORMAT_SI),
+    });
+
+    static auto Enum_Currency = ConfigEnum<sint32>(
+    {
+        ConfigEnumEntry<sint32>("IMPERIAL", MEASUREMENT_FORMAT_IMPERIAL),
```
Comment (Gymnasiast) [inline src/openrct2/config/Config.cpp]: These three entries do not belong here.

File: src/openrct2/config/KeyboardShortcuts.cpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        SDL_SCANCODE_TAB,					// SHORTCUT_SHOW_MAP
+        PLATFORM_MODIFIER | SDL_SCANCODE_S,	// SHORTCUT_SCREENSHOT
+
+        // New
+        SDL_SCANCODE_MINUS,					// SHORTCUT_REDUCE_GAME_SPEED,
+        SDL_SCANCODE_EQUALS,				// SHORTCUT_INCREASE_GAME_SPEED,
+        PLATFORM_MODIFIER | ALT | SDL_SCANCODE_C,	// SHORTCUT_OPEN_CHEAT_WINDOW,
+        SDL_SCANCODE_T,						// SHORTCUT_REMOVE_TOP_BOTTOM_TOOLBAR_TOGGLE,
+        SDL_SCANCODE_UP,					// SHORTCUT_SCROLL_MAP_UP
+        SDL_SCANCODE_LEFT,					// SHORTCUT_SCROLL_MAP_LEFT
```
Comment (Gymnasiast) [inline src/openrct2/config/KeyboardShortcuts.cpp]: This file contains tabs.

**Comment by janisozaur** [comment]
Just the Xcode now. Ping @marijnvdwerf @LRFLEW

**Comment by IntelOrca** [comment]
@janisozaur thanks

**Comment by LRFLEW** [comment]
Done. Sorry about the delay. There was a missing new config header include in the macOS platform file, so I fixed that as well. Feel free to squash that commit into the other related one.

**Comment by Gymnasiast** [comment]
There is also still the issue that KeyboardShortcuts (and possibly other files) use tabs for alignment of comments.

File: src/openrct2/scenario/scenario.h [hunk 3]
```diff
@@ -368,7 +381,7 @@ extern rct_s6_info gS6Info;
 extern char gScenarioName[64];
 extern char gScenarioDetails[256];
 extern char gScenarioCompletedBy[32];
-extern char gScenarioSavePath[MAX_PATH];
+extern char gScenarioSavePath[260];
```
Comment (janisozaur) [inline src/openrct2/scenario/scenario.h]: Why this change?

File: src/openrct2/config/ConfigEnum.h [hunk 4]
```diff
@@ -0,0 +1,79 @@
+#pragma region Copyright (c) 2014-2016 OpenRCT2 Developers
```
Comment (IntelOrca) [inline src/openrct2/config/ConfigEnum.h]: `*.hpp` is when there are implementations in the header. Although I admit in this case there is due to the template. But hpp does not mean C++ header - it means a C++ header containing C++ implementation code.

File: src/openrct2/scenario/scenario.h (hunk 3 above)
Comment (IntelOrca) [inline src/openrct2/scenario/scenario.h]: Because I don't like and causes too many issues including platform.h in header files.

File: src/openrct2/config/ConfigEnum.h (hunk 4 above)
Comment (janisozaur) [inline src/openrct2/config/ConfigEnum.h]: There is `ConfigEnum` implementation in this class, this is why I asked about it not being `.hpp`

File: src/openrct2/scenario/scenario.h (hunk 3 above)
Comment (janisozaur) [inline src/openrct2/scenario/scenario.h]: I don't like this change, Linux and other systems have larger path limits and we should start supporting it.

File: src/openrct2/config/ConfigEnum.h (hunk 4 above)
Comment (IntelOrca) [inline src/openrct2/config/ConfigEnum.h]: I'll change it, xcode will be have to be updated again

File: src/openrct2/scenario/scenario.h (hunk 3 above)
Comment (IntelOrca) [inline src/openrct2/scenario/scenario.h]: I completely agree, but platform.h is just too big and bulky to include. Consider a new header file containing the limits or use `common.h`.

File: src/openrct2/config/Config.h [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+#endif
+    extern GeneralConfiguration         gConfigGeneral;
+    extern InterfaceConfiguration       gConfigInterface;
+    extern SoundConfiguration           gConfigSound;
+    extern TwitchConfiguration          gConfigTwitch;
+    extern NetworkConfiguration         gConfigNetwork;
+    extern NotificationConfiguration    gConfigNotifications;
+    extern FontConfiguration            gConfigFonts;
+
+    bool config_open(const utf8 * path);
```
Comment (IntelOrca) [inline src/openrct2/config/Config.h]: done

File: src/openrct2/config/IniReader.cpp [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+class IniReader final : public IIniReader
+{
+private:
+    std::vector<uint8>                              _buffer;
+    std::vector<Span>                               _lines;
+    std::unordered_map<std::string, LineRange>      _sections;
+    std::unordered_map<std::string, std::string>    _values;
+
+public:
+    IniReader(const std::string &path)
```
Comment (IntelOrca) [inline src/openrct2/config/IniReader.cpp]: done

File: src/openrct2/config/IniReader.cpp [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+        return std::stof(value);
+    }
+
+    std::string GetString(const std::string &name, const std::string &defaultValue) const override
+    {
+        auto it = _values.find(name);
+        if (it == _values.end())
+        {
+            return defaultValue;
+        }
```
Comment (IntelOrca) [inline src/openrct2/config/IniReader.cpp]: done

**Comment by janisozaur** [comment]
@IntelOrca I've added some tests and I'm happy with this to be merged now (pending green CIs).

File: src/openrct2/config/IniReader.cpp [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+        return result;
+    }
+
+    sint32 GetSint32(const std::string &name, sint32 defaultValue) const override
+    {
+        sint32 result = defaultValue;
+        std::string value;
+        if (TryGetString(name, &value))
+        {
+            result = std::stoi(value);
```
Comment (janisozaur) [inline src/openrct2/config/IniReader.cpp]: Just so you know: https://github.com/OpenRCT2/OpenRCT2/pull/5240/commits/8143be0707b9252652dc9a99a28c1927054a856b#diff-e872ff6bb0551fec33f38f976677859fR48

File: test/tests/IniWriterTest.cpp [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+TEST_F(IniWriterTest, create_one_section)
+{
+    MemoryStream * ms = new MemoryStream(1000);
+    ASSERT_NE(ms, nullptr);
+    IIniWriter * iw = CreateIniWriter(ms);
+    ASSERT_NE(iw, nullptr);
+    iw->WriteSection("OpenRCT2");
+    int8_t null_terminator = 0;
```
Comment (IntelOrca) [inline test/tests/IniWriterTest.cpp]: `uint8`?

File: test/tests/IniWriterTest.cpp [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+    const char * ini = (const char *)ms->ReadString();
+    ASSERT_STREQ(ini, "path = \"C:'\\\\some/dir\\\\here/\xE7\xA5\x9E\xE9\xB7\xB9\xE6\x9A\xA2\xE9\x81\x8A\"\n");
+    Memory::Free(ini);
+    delete iw;
+    delete ms;
+}
+
+TEST_F(IniWriterTest, create_multiple_section_with_values)
+{
+    MemoryStream * ms = new MemoryStream(1000);
```
Comment (IntelOrca) [inline test/tests/IniWriterTest.cpp]: These could go on stack.

File: test/tests/IniWriterTest.cpp [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    iw->WriteSection("string");
+    iw->WriteString("path", u8"C:'\\some/dir\\here/神鷹暢遊");
+    int8_t null_terminator = 0;
+    ms->Write(&null_terminator, 1);
+    ASSERT_EQ(ms->GetPosition(), 99);
+    ASSERT_EQ(ms->GetLength(), 99);
+    ms->SetPosition(0);
+    const char * ini = (const char *)ms->ReadString();
+    ASSERT_STREQ(ini, "[bool]\nboolval = true\n\n[int]\none = 1\nzero = 0\n\n[string]\npath = "
+                      "\"C:'\\\\some/dir\\\\here/\xE7\xA5\x9E\xE9\xB7\xB9\xE6\x9A\xA2\xE9\x81\x8A\"\n");
```
Comment (IntelOrca) [inline test/tests/IniWriterTest.cpp]: Is there an assert for `std::string`? Either way you could use ms->ReadStdString() so you don't have to free the memory manually.

File: test/tests/IniReaderTest.cpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    IIniReader * ir = CreateIniReader(ms);
+    ASSERT_NE(ir, nullptr);
+    ASSERT_EQ(ir->ReadSection("doesnt_exist"), false);
+    ASSERT_EQ(ir->ReadSection("bool"), true);
+    // name of section
+    ASSERT_EQ(ir->GetSint32("bool", 42), 42);
+    // value from different section
+    ASSERT_EQ(ir->GetSint32("one", 42), 42);
+    // existing value as different type
+    EXPECT_THROW(ir->GetSint32("boolval", 42), std::invalid_argument);
```
Comment (IntelOrca) [inline test/tests/IniReaderTest.cpp]: I didn't know the convert functions throw this... this will need to be caught in Config.cpp.

File: test/tests/IniWriterTest.cpp (hunk 10 above)
Comment (janisozaur) [inline test/tests/IniWriterTest.cpp]: Oh, I did not realise.

File: test/tests/IniWriterTest.cpp (hunk 9 above)
Comment (janisozaur) [inline test/tests/IniWriterTest.cpp]: done

File: test/tests/IniWriterTest.cpp (hunk 10 above)
Comment (janisozaur) [inline test/tests/IniWriterTest.cpp]: done

File: test/tests/IniReaderTest.cpp (hunk 12 above)
Comment (janisozaur) [inline test/tests/IniReaderTest.cpp]: done

File: test/tests/StringTest.cpp [hunk 13]
```diff
@@ -0,0 +1,27 @@
+#include <string>
+#include <tuple>
+#include <gtest/gtest.h>
+#include <openrct2/core/String.hpp>
```
Comment (janisozaur) [inline test/tests/StringTest.cpp]: `"`

Comment (IntelOrca) [inline test/tests/StringTest.cpp]: no

File: src/openrct2/config/IniReader.cpp [hunk 14]
```diff
@@ -144,11 +144,8 @@ class IniReader final : public IIniReader
 
     std::string GetString(const std::string &name, const std::string &defaultValue) const override
     {
-        std::string result;
-        if (TryGetString(name, &result))
-        {
-            result = defaultValue;
-        }
+        std::string result = defaultValue;
+        TryGetString(name, &result);
```
Comment (IntelOrca) [inline src/openrct2/config/IniReader.cpp]: The code I had before was safer in case TryGetString changes `result`.

Comment (janisozaur) [inline src/openrct2/config/IniReader.cpp]: Should I change it to the old style?

Comment (IntelOrca) [inline src/openrct2/config/IniReader.cpp]: Up to you, I was just saying why I did the way it was.

**Comment by janisozaur** [comment]
I've cancelled some CI runs for this PR
